src/framework/components/zone/component.js

```diff
@@ -78,47 +102,290 @@ class ZoneComponent extends Component {
+            } else if (this.shape === 'sphere' && this.radius > 0) {
+                this._collisionShape = new Ammo.btSphereShape(this.radius);
+            }
+        }
+    }
+
+    /**
+     * Check if a point is within the zone.
+     *
+     * @param {Vec3} point - The point to look for.
+     * @param {Vec3} position - The position of this entity.
+     * @param {import('../../../core/math/quat').Quat} rotation - The rotation of this entity.
+     * @param {Mat4} matrix - The matrix for box calculations.
+     * @returns {boolean} Whether the point is within the zone.
+     * @private
+     */
+    _isPointInZone(point, position = this.entity.getPosition(), rotation = this.entity.getRotation(), matrix = undefined) {
+        if (this.shape === 'box') {
+            if (!matrix) {
+                matrix = _matrix;
+                matrix.setTRS(position, rotation, Vec3.ONE);
+                matrix.invert();
+            }
+
+            const localPoint = matrix.transformPoint(point);
+            const halfExtents = this.halfExtents;
+            if (Math.abs(localPoint.x) <= halfExtents.x && Math.abs(localPoint.y) <= halfExtents.y && Math.abs(localPoint.z) <= halfExtents.z) {
+                return true;
+            }
+        } else if (point.distance(position) <= this.radius) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Check if a point is within the zone.
+     *
+     * @param {Vec3} point - The point to look for.
+     * @returns {boolean} Whether the point is within the zone.
+     */
+    isPointInZone(point) {
+        return this._isPointInZone(point);
+    }
+
+    /**
+     * Refresh the list of entities within the zone.
+     *
+     * @ignore
+     */
+    checkEntities() {
+        if (!this.entity.enabled || !this.enabled) {
             return;
+        }
+
+        const position = this.entity.getPosition();
+        const rotation = this.entity.getRotation();
+
+        _matrix.setTRS(position, rotation, Vec3.ONE);
+        _matrix.invert();
+
+        let pendingCollider;
+        const entities = Object.values(this.system.app._entityIndex);
+
+        // Check entities as per position
+        for (let i = 0, l = entities.length; i < l; i++) {
+            const entity = entities[i];
+
+            // Don't check for self.
+            if (entity === this.entity) {
+                continue;
+            }
+
+            const index = this.entities.indexOf(entity);
+
+            if (!entity.enabled) {
+                if (index !== -1) {
+                    this.entities.splice(index, 1);
+                }
+
+                continue;
+            }
 
-        this._oldState = state;
+            if (this._isPointInZone(entity.getPosition(), position, rotation, _matrix)) {
+                if (index === -1) {
+                    entity.fire('zoneEnter', this);
+                    this.fire('entityEnter', entity);
+                    this.entities.push(entity);
```

Comment:
Is there a particular reason the "zones" list is updated after the event?  I'm currently struggling with some bugs because I was relying on this value inside of the event.  I have extra logic to determine the closest zone for when an object is inside more than one at a time e.g. a reflection probe zone where theres one large one covering the whole level and then smaller ones inside of it that may overlap each other for each area.

Reply:
Thank you for the feedback. I changed the order and made the documentation a bit more clear regarding the order!

Reply:
nice one.  This is great work